### PR TITLE
removing extra item on initial load in for blog index page

### PR DESCRIPTION
**Pull request by danielfmiranda**
Related PRs/issues: #9190
# Description

In this PR I introduced a if statement on the BlogIndexPage model, that will remove the extra item from the initial entry set. 

The if statement checks:
- If the page is filtered (meaning we wouldnt be on the landing page with the callout box, meaning no update is necessary)
- If the page size is set to 12 or 24 (the page sizes where this problem is occuring)

I only check for 12 or 24 because these are the page sizes that are divisible by 3, which is what the columns are set to. 4 and 8 always have extra items on the row, though the page size of 8 loads in nicely since the empty space is taken by the callout box. 

# Screenshots:

 **Screenshot of bug:**

Screen Shot 2022-08-08 at 1 09 22 PM

**Page Size 4:**

Note that the initial entry count is unaffected, as page size of 4 or 8 are going to have overflowing cards anyway.
Screen Shot 2022-08-08 at 1 09 22 PM

Screen Shot 2022-08-08 at 1 09 22 PM

**Page Size 12:**

Note that the initial entry count is - 1 so there is no extra card, and then loading more cards renders 4 rows as expected
Screen Shot 2022-08-08 at 1 09 22 PM
Screen Shot 2022-08-08 at 1 09 22 PM

# Checklist

**Tests**
- [ ] ~~Is the code I'm adding covered by tests?~~

**Changes in Models:**
- [ ] ~~Did I update or add new fake data?~~
- [ ] ~~Did I squash my migration?~~
- [ ] ~~Are my changes [backward-compatible](https://github.com/mozilla/foundation.mozilla.org/blob/main/docs/workflow.md#django-migrations-what-to-do-when-working-on-backward-incompatible-migrations). If not, did I schedule a deploy with the rest of the team?~~

**Documentation:**
- [ ] ~~Is my code documented?~~
- [ ] ~~Did I update the READMEs or wagtail documentation?~~

### CODE REVIEW

File: network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py [hunk 1]
```diff
@@ -155,6 +155,13 @@ class BlogIndexPage(IndexPage):
     def get_context(self, request):
         context = super().get_context(request)
         context["related_topics"] = self.get_related_topics()
+
+        if not hasattr(self, 'filtered') and self.page_size == 12 or self.page_size == 24:
```
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:159]: Making sure we are not on a filtered page, which do not render the callout box at all.

Also making sure that the page size is 12/24, which are the only page sizes divisible by 3 and having this issue.

File: network-api/networkapi/wagtailpages/pagemodels/index.py [hunk 2]
```diff
@@ -202,6 +203,9 @@ def generate_entries_set_html(self, request, *args, **kwargs):
                 pass
 
         start = page * page_size
+        if 'entries_offset_for_callout' in context:
+            start = start - 1
```
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/index.py:207]: If the BlogIndexPages initial entries were offset by the callout box (for example setting initial entry list size from 12 -> 11), we then need to update the starting point of the next page.

If we dont, the next page is going to start at 13, leaving entry #12 out of the rendered list.

Comment (tbrlpld) [reply network-api/networkapi/wagtailpages/pagemodels/index.py:207]: **Question:** Do we use callout boxes in other index pages than the blog index? If not, we should not make this change on the index page, but rather keep it to the blog index.

Comment (tbrlpld) [reply network-api/networkapi/wagtailpages/pagemodels/index.py:207]: Moving this to the blog index would also mean you don't need to pull in the context that you don't really need here.

File: network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py (hunk 1 above)
Comment (tbrlpld) [reply network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:159]: It would be good to move this check into a helper method that clearly states what this check is supposed to mean. Something like `is_showing_callout_box_in_entries`. 

Also, we probably don't want to show the callout box if there are not topics.

File: network-api/networkapi/wagtailpages/pagemodels/index.py (hunk 2 above)
Comment (tbrlpld) [reply network-api/networkapi/wagtailpages/pagemodels/index.py:207]: I think it would be good to add some tests around this to cover all the edge cases, but I think in general this approach (of adjusting the start point) is fine.

**Review by tbrlpld — DISMISSED** [review]

File: network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py (hunk 1 above)
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:159]: Great Idea! Updated and also added the check for the topics

File: network-api/networkapi/wagtailpages/templates/wagtailpages/blog_index_page.html [hunk 3]
```diff
@@ -75,7 +75,7 @@ <h2 class="tw-hidden tw-uppercase tw-text-3xl tw-mb-0 tw-font-bold tw-font-sans
     <ul id="index-entries" class="tw-grid medium:tw-grid-cols-2 large:tw-grid-cols-3 tw-gap-x-5 tw-gap-y-4 tw-list-none tw-divide-y tw-divide-gray-20 tw-px-[15px] medium:tw-divide-y-0">
       {% include "./fragments/entry_cards_item_loop.html" %}   
 
-      {% if not filtered %}
+      {% if not filtered and related_topics %}
```
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/templates/wagtailpages/blog_index_page.html:78]: If there are no topics set, do not render the topics box

File: network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py [hunk 4]
```diff
@@ -306,6 +313,68 @@ def extract_topic_information(self, topic_slug):
             'topic': topic_object
         }
 
+    def get_related_topics(self):
+        related_topics = self.related_topics.all()
+        return related_topics
+
+    # Helper function to show if entry list is offset by topic box
+    def is_showing_topics_box_in_entry_list(self):
```
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:321]: New helper function as suggested from feedback

File: network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py [hunk 5]
```diff
@@ -306,6 +313,68 @@ def extract_topic_information(self, topic_slug):
             'topic': topic_object
         }
 
+    def get_related_topics(self):
```
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:316]: Moved this topic up here because I thought it would make sense to have all the routes at the end of the file.

File: network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py [hunk 6]
```diff
@@ -306,6 +313,68 @@ def extract_topic_information(self, topic_slug):
...
+        if (
+            not hasattr(self, "filtered")
+            and self.get_related_topics().count()
+            and (self.page_size == 12 or self.page_size == 24)
+        ):
+            return True
+        else:
+            return False
+
+    @route('^entries/')
```
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:331]: Per request from feedback, I have moved the function for rendering new entries here, and simplified a little bit by removing any unneeded logic + adding the check for whether or not we need to update the index by 1.

**Comment by danielfmiranda** [comment]
Thanks for the great feedback @tbrlpld!

Per request I have:
- Added some tests
- Created a helper function for checking whether or not the list is offset
- Updated blog index to load its new entry pages, instead of relying on index.py's

**Review by fessehaye — APPROVED** [review]
tests look good to me
